sklearn/preprocessing/imputation.py

```diff
@@ -390,3 +403,538 @@ def transform(self, X):
+            self.random_state_.shuffle(ordered_idx)
+        else:
+            raise ValueError("Got an invalid imputation order: '{0}'. It must "
+                             "be one of the following: 'roman', 'arabic', "
+                             " 'monotone', 'revmonotone', or "
+                             "'random'.".format(self.imputation_order))
+        return ordered_idx
+
+    def _get_abs_corr_mat(self, X_filled, tolerance=1e-6):
+        """Gets absolute correlation matrix between features.
+
+        Parameters
+        ----------
+        X_filled : array-like, shape (n_samples, n_features)
+            Input data with the most recent imputations.
+
+        tolerance : float, optional (default=1e-6)
+            `abs_corr_mat` can have nans, which will be replaced
+            with `tolerance`.
+
+        Returns
+        -------
+        abs_corr_mat : array-like, shape (n_features, n_features)
+            Absolute correlation matrix of X at the beginning of the current
+            round. The diagonal has been zeroed out and each feature's absolute
+            correlations with all others have been normalized to sum to 1.
+        """
+        n_features = X_filled.shape[1]
+        if (self.n_nearest_features is None or
+                self.n_nearest_features >= n_features):
+            return None
+        abs_corr_mat = np.abs(np.corrcoef(X_filled.T))
+        # np.corrcoef is not defined for features with zero std
+        abs_corr_mat[np.isnan(abs_corr_mat)] = tolerance
+        # ensures exploration, i.e. at least some probability of sampling
+        abs_corr_mat[abs_corr_mat < tolerance] = tolerance
+        # features are not their own neighbors
+        np.fill_diagonal(abs_corr_mat, 0)
+        # needs to sum to 1 for np.random.choice sampling
+        abs_corr_mat = normalize(abs_corr_mat, norm='l1', axis=0, copy=False)
+        return abs_corr_mat
+
+    def _initial_imputation(self, X):
+        """Performs initial imputation for input X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Input data, where "n_samples" is the number of samples and
+            "n_features" is the number of features.
+
+        Returns
+        -------
+        Xt : array-like, shape (n_samples, n_features)
+            Input data, where "n_samples" is the number of samples and
+            "n_features" is the number of features.
+
+        X_filled : array-like, shape (n_samples, n_features)
+            Input data with the most recent imputations.
+
+        mask_missing_values : array-like, shape (n_samples, n_features)
+            Input data's missing indicator matrix, where "n_samples" is the
+            number of samples and "n_features" is the number of features.
+        """
+        faf = 'allow-nan' if self.missing_values == 'NaN' else True
+        X = check_array(X, dtype=np.float32, order="F", force_all_finite=faf)
```

Comment:
I would put the faf in the check_array

Reply:
any reason to force float 32. Can we accept bother 32, 64 without making a copy?

Reply:
I just copied what was in `Imputer`, assuming they had a good reason to do what they did. I can instead do `FLOAT_DTYPES`?